core/capability-negotiation.md

```diff
@@ -0,0 +1,457 @@
+
+    Client: CAP LS 302
+    Server: CAP * LS * :multi-prefix extended-join account-notify batch invite-notify tls
+    Server: CAP * LS * :cap-notify server-time example.org/dummy-cap=dummyvalue example.org/second-dummy-cap
+    Server: CAP * LS :userhost-in-names sasl=EXTERNAL,DH-AES,DH-BLOWFISH,ECDSA-NIST256P-CHALLENGE,PLAIN
+
+Example of a `LIST` reply with a client that supports CAP version `302`:
+
+    Client: CAP LIST
+    Server: CAP modernclient LIST * :example.org/example-cap example.org/second-example-cap account-notify
+    Server: CAP modernclient LIST :invite-notify batch example.org/third-example-cap
+
+### The CAP LIST subcommand
+
+The LIST subcommand is used to list the capabilities associated with the active connection.
+The client should send a LIST subcommand with no other arguments to solicit a list of
+active capabilities.
+
+If no capabilities are active, an empty parameter must be sent.
+
+Example:
+
+    Client: CAP LIST
+    Server: CAP * LIST :multi-prefix
+
+### The CAP REQ subcommand
+
+The REQ subcommand is used to request a change in capabilities associated with the active
+connection.  Its sole parameter must be a list of space-separated capability identifiers.
+Each capability identifier may be prefixed with a dash (`-`) to designate that the capability
+should be disabled.
+
+If a client requests a capability which is already enabled, or tries to disable a capability
+which is not enabled, the server MUST continue processing the REQ subcommand as though
+handling this capability was successful.
+
+The capability identifier set must be accepted as a whole, or rejected entirely.
+
+If a client issues a REQ subcommand, registration must be suspended until an END subcommand
+is received.
+
+Example adding a capability:
+
+    Client: CAP REQ :multi-prefix sasl
+    Server: CAP * ACK :multi-prefix sasl
+
+Example removing a capability:
+
+    Client: CAP REQ :-userhost-in-names
+    Server: CAP * ACK :-userhost-in-names
+
+### The CAP ACK subcommand
+
+The ACK subcommand is sent by the server to acknowledge a client-sent REQ, and let the client
+know that their requested capabilities have been enabled.
+
+This subcommand has a single parameter consisting of a space-separated list of capability
+names. Each capability name may be prefixed with a dash (`-`), indicating that this
+capability has been disabled as requested.
+
+If an ACK reply originating from the server is spread across multiple lines, a client MUST NOT
+change capabilities until the last ACK of the set is received. Equally, a server MUST NOT
+change the capabilities of the client until the last ACK of the set has been sent.
+
+### The CAP NAK subcommand
+
+The NAK subcommand designates that the requested capability change was rejected.  The server
+MUST NOT make any change to any capabilities if it replies with a NAK subcommand.
+
+The argument of the NAK subcommand MUST consist of at least the first 100 characters of the
```

Comment:
Anyone remember why 100? I guess a client could send a really long (+512) list of REQs, but not sure why we need to define specific additional ytruncation rules here.

Reply:
No idea why 100. It goes all the way back to the 2005 [mitchell ietf draft][1]. The 2002 and 2003 drafts don't mention anything like this.

[1]: https://tools.ietf.org/html/draft-mitchell-irc-capabilities-00#section-3.5

The changelog entry `2005-01-21 KLM` says

```
      *  Explicitly note in the ABNF for NAK that the parameter is an
         opaque repeat of at least the first 100 characters of the
         argument to REQ.
```

The original text:

```
   [...]
   subcommand.  The argument of the NAK subcommand MUST consist of at
   least the first one hundred characters of the capability list in the
   REQ (Section 3.3) subcommand which triggered the NAK.

   ABNF [4] description of the NAK subcommand:

          nakcmd       =  "NAK" SP ":" acklist
                           ; acklist is at least 100 characters of the
                           ; capability list from the REQ
```

Reply:
this feels weird. I guess it's explicitly mentioned for... reasons of making sure the client can match up its' REQs to NAKs, but... yeah it feels strange. I can go either way on it, unless we get some more info from someone that actually uses this explicitly?